packages/mdc-tooltip/foundation.js

```diff
@@ -0,0 +1,173 @@
+ * limitations under the License.
+ */
+
+import MDCFoundation from '@material/base/foundation';
+import MDCTooltipAdapter from './adapter';
+import cssClasses from './constants';
+
+/**
+ * @extends {MDCFoundation<!MDCTooltipAdapter>}
+ */
+class MDCTooltipFoundation extends MDCFoundation {
+  /** @return enum {cssClasses} */
+  static get cssClasses() {
+    return cssClasses;
+  }
+
+  /** @return {!MDCTooltipAdapter} */
+  static get defaultAdapter() {
+    return /** @type {!MDCTooltipAdapter} */ ({
+      addClass: (/* className: string */) => {},
+      removeClass: (/* className: string */) => {},
+      getClassList: () => /* [classNames: string] */ [],
+      getRootWidth: () => /* type: number */ 0,
+      getRootHeight: () => /* type: number */ 0,
+      getControllerWidth: () => /* type: number */ 0,
+      getControllerHeight: () => /* type: number */ 0,
+      getControllerOffsetTop: () => /* type: number */ 0,
+      getControllerOffsetLeft: () => /* type: number */ 0,
+      setStyle: (/* propertyName: string, value: string */) => {},
+      registerListener: (/* type: string, handler: EventListener */) => {},
+      deregisterListener: (/* type: string, handler: EventListener */) => {}
+    });
+  }
+
+  constructor(adapter) {
+    super(Object.assign(MDCTooltipFoundation.defaultAdapter, adapter));
+
+    /** @private {boolean} */
+    this.displayed_ = false;
+
+    /** @private {?string} */
+    this.direction_ = null;
+
+    /** @private {?number} */
+    this.showTimeout_ = null;
+
+    /** @private {?number} */
+    this.hideTimeout_ = null;
+
+    /** @public {number} */
+    this.gap = 12;
+
+    /** @public {number} */
+    this.showDelay = 0;
+
+    /** @public {number} */
+    this.hideDelay = 1500;
+  }
+
+  init() {
+    this.addEventListeners_();
+  }
+
+  setDirection_() {
+    this.direction_ = 'bottom';
+    const possibleDirections = {};
+    possibleDirections[cssClasses.DIRECTION_BOTTOM] = 'bottom';
+    possibleDirections[cssClasses.DIRECTION_TOP] = 'top';
+    possibleDirections[cssClasses.DIRECTION_LEFT] = 'left';
+    possibleDirections[cssClasses.DIRECTION_RIGHT] = 'right';
+
+    const classNames = this.adapter_.getClassList();
+
+    for (let i = 0; i < classNames.length; i++) {
+      if (possibleDirections[classNames[i]] !== undefined) {
+        this.direction_ = possibleDirections[classNames[i]];
+      }
+    }
+
+    const tooltipHeight = this.adapter_.getRootHeight();
+    const tooltipWidth = this.adapter_.getRootWidth();
+    const ctrlOffsetTop = this.adapter_.getControllerOffsetTop();
+    const ctrlOffsetLeft = this.adapter_.getControllerOffsetLeft();
+    const ctrlHeight = this.adapter_.getControllerHeight();
+    const ctrlWidth = this.adapter_.getControllerWidth();
+
+    let top = ctrlOffsetTop + ctrlHeight / 2 - tooltipHeight / 2;
+    let left = ctrlOffsetLeft + ctrlWidth / 2 - tooltipWidth / 2;
+
+    if (this.direction_ === 'top') {
```

Comment:
Still I'm seeing the tooltip being cutoff in smaller screen resolutions
![screenshot from 2018-07-06 16-12-32](https://user-images.githubusercontent.com/579873/42403697-d62736b0-8137-11e8-8ce8-c91fc3d93f5d.png)

It looks like there is enough space to the right, but its still cut off. I'm not sure what the correct behavior in this case would be. Maybe the gap should decrease? I'll have to check with design.

Reply:
I have not yet implemented the detection if a tooltip is moving outside of the viewport.